cardano-crypto-tests/src/Bench/Crypto/KES.hs

```diff
@@ -2,62 +2,92 @@
 {-#LANGUAGE ScopedTypeVariables #-}
 {-#LANGUAGE TypeFamilies #-}
 {-#LANGUAGE FlexibleContexts #-}
+{-#LANGUAGE OverloadedStrings #-}
+{-#LANGUAGE PolyKinds #-}
+{-#LANGUAGE RankNTypes #-}
 module Bench.Crypto.KES
   ( benchmarks
   ) where
 
 import Data.Proxy
 import Data.Maybe (fromJust)
-import Data.ByteString (ByteString)
 
 import Control.DeepSeq
 
-import Cardano.Crypto.DSIGN.Ed25519
+import Cardano.Crypto.DSIGN.Ed25519ML
 import Cardano.Crypto.Hash.Blake2b
 import Cardano.Crypto.KES.Class
 import Cardano.Crypto.KES.Sum
 import Cardano.Crypto.KES.CompactSum
 
+import Control.Concurrent.MVar
+import Control.Exception (bracket)
+
 import Criterion
+import qualified Data.ByteString as BS (ByteString)
+import Data.Either (fromRight)
+import Cardano.Crypto.Libsodium as NaCl
+import Cardano.Crypto.MLockedSeed
+import System.IO.Unsafe (unsafePerformIO)
+import GHC.TypeLits (KnownNat)
+import Data.Kind (Type)
 
 import Bench.Crypto.BenchData
 
+{- HLINT ignore "Use camelCase" -}
+
+{-# NOINLINE testSeedML #-}
+testSeedML :: forall n. KnownNat n => MLockedSeed n
+testSeedML = MLockedSeed . unsafePerformIO $ NaCl.mlsbFromByteString testBytes
+
+withKESLock :: MVar () -> IO a -> IO a
+withKESLock kesLock action =
+  bracket
+    (takeMVar kesLock)
+    (putMVar kesLock)
+    (const action)
 
-benchmarks :: Benchmark
-benchmarks = bgroup "KES"
-  [ benchKES @Proxy @(Sum6KES Ed25519DSIGN Blake2b_256) Proxy "Sum6KES"
-  , benchKES @Proxy @(CompactSum6KES Ed25519DSIGN Blake2b_256) Proxy "CompactSum6KES"
+benchmarks :: MVar () -> Benchmark
+benchmarks kesLock = bgroup "KES"
+  [ benchKES @Proxy @(Sum6KES Ed25519DSIGNM Blake2b_256) Proxy kesLock "Sum6KES"
+  , benchKES @Proxy @(Sum7KES Ed25519DSIGNM Blake2b_256) Proxy kesLock "Sum7KES"
+  , benchKES @Proxy @(CompactSum6KES Ed25519DSIGNM Blake2b_256) Proxy kesLock "CompactSum6KES"
+  , benchKES @Proxy @(CompactSum7KES Ed25519DSIGNM Blake2b_256) Proxy kesLock "CompactSum7KES"
   ]
 
-benchKES :: forall proxy v
-          . ( KESAlgorithm v
-            , ContextKES v ~ ()
-            , Signable v ByteString
-            , NFData (SignKeyKES v)
-            , NFData (VerKeyKES v)
-            , NFData (SigKES v)
-            )
-         => proxy v
-         -> [Char]
-         -> Benchmark
-benchKES _ lbl =
+
+
+{-# NOINLINE benchKES #-}
+benchKES :: forall (proxy :: forall k. k -> Type) v
+           . ( KESSignAlgorithm IO v
+             , ContextKES v ~ ()
+             , Signable v BS.ByteString
+             , NFData (SignKeyKES v)
+             , NFData (SigKES v)
+             )
+          => proxy v
+          -> MVar ()
+          -> [Char]
+          -> Benchmark
+benchKES _ kesLock lbl =
```

Comment:
Why do you need a semaphore for benchmarks?

Regardless of the fact that KES implementation is thread safe criterion benchmarks aren't executed in parallel. They are all done sequentially,  so there is no chance of concurrent access. I think all of this lock business needs to go.

Reply:
Yes, see above.

Reply:
The thought behind the lock wasn't thread safety btw., but avoiding excessive mlocking. At the time, running all those property tests concurrently would fail, so I added the lock to force them to run sequentially. However, with the pool allocator, it looks like that's no longer necessary. I'll have to verify on CI though; my own machine proves to not be a very good predictor of this, though trace-logging the memory pool's page allocations seems to suggest that it will be fine.